src/main/joyride/nrepl.cljs

```diff
@@ -182,51 +184,102 @@
     ;; Expose "app" key under js/app in the repl
-    (.listen server
-             port
-             "127.0.0.1" ;; default for now
-             (fn []
-               (let [addr (-> server (.address))
-                     port (-> addr .-port)
-                     host (-> addr .-address)]
-                 (info "nREPL server started on port" port "on host"
-                       (str host "- nrepl://" host ":" port))
-                 (->
-                   (vscode/workspace.fs.writeFile (vscode/Uri.file
-                                                   (path/join vscode/workspace.rootPath ".joyride" ".nrepl-port"))
-                                                  (-> (new js/TextEncoder) (.encode (str port))))
-                   (p/catch
-                       (fn [e]
-                         (info "Could not write .nrepl-port" e)))))))
-    server)
-  #_(let [onExit (js/require "signal-exit")]
-      (onExit (fn [_code _signal]
-                (debug "Process exit, removing .nrepl-port")
-                (fs/unlinkSync ".nrepl-port")))))
-
-(defn stop-server [server]
-  (.close server
-          (fn []
-            (when (fs/existsSync ".nrepl-port")
-              (fs/unlinkSync ".nrepl-port")))))
+
+    #_(on-exit (fn [_code _signal]
+                 (debug "Process exit, removing port file")
+                 (remove-port-file (port-file-uri))))
+
+    (p/create (fn [resolve reject]
+                (try
+                  (let [server (node-net/createServer
+                                (partial on-connect {:sci-ctx-atom ctx-atom
+                                                     :sci-last-error sci-last-error}))]
+                    (swap! !db assoc ::server server)
+                    (when-contexts/set-context! ::when-contexts/joyride.isNReplServerRunning true)
+                    (.listen server
+                             port
+                             "127.0.0.1" ;; default for now
+                             (fn []
+                               (let [addr (-> server (.address))
+                                     port (-> addr .-port)
+                                     host (-> addr .-address)]
+                                 (info "nREPL server started on port" port "on host"
+                                       (str host "- nrepl://" host ":" port))
+                                 (-> (vscode/workspace.fs.writeFile (port-file-uri root-path)
+                                                                    (-> (new js/TextEncoder) (.encode (str port))))
+                                     (p/handle (fn [_result error]
+                                                 (resolve port)
+                                                 (when error
+                                                   (info "Could not write port file" error)))))))))
+                  (catch :default e
+                    (reject (ex-info "Error creating nREPL server" e))))))))
```

Comment:
@borkdude please help a noob here. I want this to be a ”proper” error, that the user can catch something like:

```clojure
  ;; Starting the nREPL server
  (-> (.startNReplServer joyApi)
      (p/catch (fn [e] (println "Booo!" (.-message e) e))))
```

Not sure how I create that...

Reply:
When you return a promise and an error occurs while the promise happens, then the user can already use `.catch` on the promise. That's the fun thing of promises, you don't have to do that yourself.

Reply:
So I should just skip doing the catch->reject? ... I'll try that.

Reply:
Yes